packages/devtools_app/lib/src/ui/search.dart

```diff
@@ -606,6 +609,7 @@ mixin SearchFieldMixin<T extends StatefulWidget> on State<T> {
     super.dispose();
     searchTextFieldController?.dispose();
     _searchFieldFocusNode?.dispose();
+    _rawKeyboardFocusNode?.dispose();
```

Comment:
why not use autoDisposeFocusNode instead of manually disposing?

Reply:
Hmm can I use a `mixin` in a `mixin` (ie `AutoDisposeMixin` within the `SearchFieldMixin`)? I'm not finding documentation about that.

Reply:
You should be able to add `AutoDisposeMixin` to the `on` list:
`mixin SearchFieldMixin<T extends StatefulWidget> on State<T>, AutoDisposeMixin<T>`

Reply:
Hmm I thought that restricted what classes could use the `mixin` (eg, `SearchFieldMixin<T>` can only be used on classes that implement both `State<T>` and `AutoDisposeMixin<T>`?

I'm seeing the following error on `FileSearchFieldState` (and other places): 

'SearchFieldMixin<StatefulWidget>' can't be mixed onto 'State<FileSearchField>' because 'State<FileSearchField>' doesn't implement 'AutoDisposeMixin<StatefulWidget>'.
Try extending the class 'SearchFieldMixin<StatefulWidget>'.

Reply:
If the `SearchFieldMixin` depends on functionality from `AutoDisposeMixin`, I think it is fine to force consumers of `SearchFieldMixin` to mixin both `SearchFieldMixin` and `AutoDisposeMixin`.

Reply:
CC @leafpetersen @munificent Is there a best practice around mixins depending on other mixins?

Reply:
cc @lrhn @natebosch for thoughts on best practices here.

Reply:
I'm not aware of any official or semi-official best practices.
I'd personally have no problem making one mixin depend on another, in order to force you to always include the other.

You can do that using the `on`, which requires the `AutoDisposeMixin` to be mixed in first. That's probably the safest approach, and it's very explicit about not being allowed to mix in `SearchFieldMixin` without first mixing in `AutoDisposeMixin`. (Not amazing, though, "Try extending the class 'SearchFieldMixin'" isn't really helping.)

You can also let `SearchFieldMixin` do `implements AutoDisposeMixin`, without actually implementing the members. That means that the user must mix in `AutoDisposeMixin` too (or implement its interface manually) in order to satisfy the interface, but the error message are not going to be as directly actionable.


Anyway, why is `SearchFieldMixin` a mixin to begin with, and not just a `Widget`. This feels like something where you'd want composition rather than inheritance, so a widget shouldn't be a search field *as well* as something else, it should be (or have) the something else and also *have* a search field. (Can something have two search fields?)

The two uses of `SearchFieldMixin` I see in internal code are using it as `extends State<Something> with SearchFieldMixin, AutoDisposeMixin`, so they would be equally well helped by having `class SearchFieldMixin<T extends Widget> extends State<T> with AutoDisposeMixin { ... }` and extending that. 
I'm not seeing the advantage of being a `mixin` here.

Reply:
> why is SearchFieldMixin a mixin to begin with, and not just a Widget

It's been a while since this code was written, but I believe this was made a mixin so that the State classes that mixin SearchFieldMixin can call these three methods from the mixin: https://github.com/flutter/devtools/blob/136617a5e3f44e6837f2ebb4ee137f46d0f236c3/packages/devtools_app/lib/src/ui/search.dart#L704-L735. These may be able to be pushed into the controller, but this would require some extra plumbing to have the text field controller respond to events from the business logic controller, which may be a little messy.

Either way, it sounds like it is fine for now to force classes mixing in SearchFieldMixin to also mixin AutoDisposeMixin.

Reply:
+1 on moving towards composition for this design. FWIW I find most designs with an imposed `with` or `extends` hard to follow especially when they mix methods which are supposed to be called by others (`initState` and `dispose` here I think) alongside methods that are supposed to be called by the extending class.

Reply:
Thanks everyone for clarifying! I've added a TODO for refactoring `SearchFieldMixin`. It looks like we currently can't use the `AutoDisposeMixin` with it due to complications with how the `AutoDisposeMixin` is structured (opened https://github.com/flutter/devtools/issues/3538 for that). Will address both in separate PRs to keep the scope of this change limited to `FocusNodes`.